lib/pos_mempool.go

```diff
@@ -0,0 +1,417 @@
+	// First, sanity check our reserved balance.
+	userPk := NewPublicKey(txn.Tx.PublicKey)
+
+	// Remove the transaction from the register.
+	if err := dmp.txnRegister.RemoveTransaction(txn); err != nil {
+		return errors.Wrapf(err, "PosMempool.removeTransactionNoLock: Problem removing txn from register")
+	}
+	// Decrease the appropriate ledger's balance by the transaction fee.
+	dmp.ledger.DecreaseBalance(*userPk, txn.Fee)
+
+	// Emit an event for the removed transaction.
+	if persistToDb {
+		event := &MempoolEvent{
+			Txn:  txn,
+			Type: MempoolEventRemove,
+		}
+		dmp.persister.EnqueueEvent(event)
+	}
+
+	return nil
+}
+
+// GetTransaction returns the transaction with the given hash if it exists in the mempool. This function is thread-safe.
+func (dmp *PosMempool) GetTransaction(txnHash *BlockHash) *MsgDeSoTxn {
+	if !dmp.IsRunning() {
+		return nil
+	}
+
+	dmp.RLock()
+	defer dmp.RUnlock()
+
+	txn := dmp.txnRegister.GetTransaction(txnHash)
+	if txn == nil || txn.Tx == nil {
+		return nil
+	}
+
+	return txn.Tx
+}
+
+// GetTransactions returns all transactions in the mempool ordered by the Fee-Time algorithm. This function is thread-safe.
+func (dmp *PosMempool) GetTransactions() []*MsgDeSoTxn {
+	if !dmp.IsRunning() {
+		return nil
+	}
+
+	dmp.RLock()
+	defer dmp.RUnlock()
+
+	var desoTxns []*MsgDeSoTxn
+	poolTxns := dmp.txnRegister.GetFeeTimeTransactions()
+	for _, txn := range poolTxns {
+		if txn == nil || txn.Tx == nil {
+			continue
+		}
+		desoTxns = append(desoTxns, txn.Tx)
+	}
+	return desoTxns
+}
+
+// GetIterator returns an iterator for the mempool transactions. The iterator can be used to peek transactions in the
+// mempool ordered by the Fee-Time algorithm. Transactions can be fetched with the following pattern:
+//
+//	for it.Next() {
+//		if txn, ok := it.Value(); ok {
+//			// Do something with txn.
+//		}
+//	}
+//
+// Note that the iteration pattern is not thread-safe. Another lock should be used to ensure thread-safety.
+func (dmp *PosMempool) GetIterator() MempoolIterator {
+	if !dmp.IsRunning() {
+		return nil
+	}
+
+	dmp.RLock()
+	defer dmp.RUnlock()
+
+	return NewPosMempoolIterator(dmp.txnRegister.GetFeeTimeIterator())
+}
+
+// pruneNoLock removes transactions from the mempool until the mempool size is below the maximum allowed size. The transactions
+// are removed in lowest to highest Fee-Time priority, i.e. opposite way that transactions are ordered in
+// GetTransactions().
+func (dmp *PosMempool) pruneNoLock() error {
+	if dmp.txnRegister.Size() < dmp.params.MaxMempoolPosSizeBytes {
+		return nil
+	}
+
+	prunedTxns, err := dmp.txnRegister.PruneToSize(dmp.params.MaxMempoolPosSizeBytes)
+	if err != nil {
+		return errors.Wrapf(err, "PosMempool.pruneNoLock: Problem pruning mempool")
+	}
+	for _, prunedTxn := range prunedTxns {
+		if err := dmp.removeTransactionNoLock(prunedTxn, true); err != nil {
+			glog.Errorf("PosMempool.pruneNoLock: Problem removing transaction from mempool: %v", err)
```

Comment:
Should we return the error here instead of glogging it?

Reply:
I guess the assumption here is that this loop will never error because the txn no longer exists in the transaction register, in which case the `dmp.removeTransactionNoLock()` call will never error.

@AeonSw4n, IMO it's fine to replace the `dmp.removeTransactionNoLock()` call with a helper function that just does the following:
```
	// Decrease the appropriate ledger's balance by the transaction fee.
	dmp.ledger.DecreaseBalance(*userPk, txn.Fee)

	// Emit an event for the removed transaction.
	if persistToDb {
		event := &MempoolEvent{
			Txn:  txn,
			Type: MempoolEventRemove,
		}
		dmp.persister.EnqueueEvent(event)
	}
```

This would guarantee there's no error returned, and clarify for us in the future how pruneNoLock() differs from removeTransactionNoLock().

Reply:
Yes, this glog should never happen, except maybe if `prunedTxn` or `prunedTxn.Hash` were nil. Either way, it's mostly a dummy error handler.

@tholonious re the code split, I prefer making the call to `removeTransactionNoLock`. It's nice that both "removeTransaction" and "removeTransactioNoLock" erase transactions. This property will not hold for the new "sub-remove" function. 

Imo it's also intuitive to call an internal remove function for each pruned txn. Perhaps making assumptions that `dmp.removeTransactionNoLock()` doesn't panic on non-existing transactions isn't that bad?

Reply:
@AeonSw4n, that seems fine.  The implementation of `dmp.removeTransactionNoLock()` is encapsulated within the same struct anyway so that's a reasonable assumption. 

Definitely add a comment in the code here that this function call will never error so it's clearer to readers